commit d6462da5dde7d38604bd5783e609ec8392eaf74c
Author: Michael Maitland <michaeltmaitland@gmail.com>
Date:   Thu Mar 20 08:21:12 2025 -0700

    fixup! clang-format

diff --git a/llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp b/llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
index 61f43e2e72b1..80ae0fd3d930 100644
--- a/llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
+++ b/llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
@@ -1342,8 +1342,8 @@ RISCVVLOptimizer::checkUsers(const MachineInstr &MI) const {
 
     if (UserMI.isPHI()) {
       LLVM_DEBUG(dbgs() << "    Peeking through uses of PHI\n");
-        for (auto &PhiUse : MRI->use_operands(UserMI.getOperand(0).getReg()))
-          Worklist.insert(&PhiUse);
+      for (auto &PhiUse : MRI->use_operands(UserMI.getOperand(0).getReg()))
+        Worklist.insert(&PhiUse);
       continue;
     }
 
