// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
// RUN: %clang_cc1  -triple aarch64-none-linux-gnu -target-feature +sve \
// RUN:   -target-feature +sme -target-feature +sve2 -target-feature +neon -fsyntax-only -verify %s

// REQUIRES: aarch64-registered-target

#include "arm_neon.h"

int16x8_t incompat_neon_sm(int16x8_t splat) __arm_streaming {
  // expected-warning@+1 {{builtin call has undefined behaviour when called from a streaming function}}
  return (int16x8_t)__builtin_neon_vqaddq_v((int8x16_t)splat, (int8x16_t)splat, 33);
}

__arm_locally_streaming int16x8_t incompat_neon_ls(int16x8_t splat) {
  // expected-warning@+1 {{builtin call has undefined behaviour when called from a streaming function}}
  return (int16x8_t)__builtin_neon_vqaddq_v((int8x16_t)splat, (int8x16_t)splat, 33);
}

int16x8_t incompat_neon_smc(int16x8_t splat) __arm_streaming_compatible {
  // expected-warning@+1 {{builtin call has undefined behaviour when called from a streaming compatible function}}
  return (int16x8_t)__builtin_neon_vqaddq_v((int8x16_t)splat, (int8x16_t)splat, 33);
}
