// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
// RUN: %clang_cc1 -triple riscv64 -target-feature +zbkb -verify %s -o -

int zip(int rs1)
{
  return __builtin_riscv_zip_32(rs1); // expected-error {{builtin requires: 'RV32'}}
}

int unzip(int rs1)
{
  return __builtin_riscv_unzip_32(rs1); // expected-error {{builtin requires: 'RV32'}}
}
