// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 5
// REQUIRES: amdgpu-registered-target
// RUN: %clang_cc1 -verify -cl-std=CL1.2 -triple amdgcn-amd-amdhsa %s
// RUN: %clang_cc1 -verify -cl-std=CL2.0 -triple amdgcn-amd-amdhsa %s

void f() {
    int n = 3;
    __amdgpu_image_rsrc_t v = 0; // expected-error {{initializing '__private __amdgpu_image_rsrc_t' with an expression of incompatible type 'int'}}
    int k = v;                   // expected-error {{initializing '__private int' with an expression of incompatible type '__private __amdgpu_image_rsrc_t'}}
    (void)(v + v);               // expected-error {{invalid operands}}
    __amdgpu_image_rsrc_t r;
    int *p = (int*)r;            // expected-error {{operand of type '__amdgpu_image_rsrc_t' where arithmetic or pointer type is required}}
    (void)p;
}
