diff --git a/clang/cmake/caches/BOLT-PGO.cmake b/clang/cmake/caches/BOLT-PGO.cmake
index 1a04ca9a74e5..d092820e4115 100644
--- a/clang/cmake/caches/BOLT-PGO.cmake
+++ b/clang/cmake/caches/BOLT-PGO.cmake
@@ -4,6 +4,8 @@ set(CLANG_BOOTSTRAP_TARGETS
   stage2-clang-bolt
   stage2-distribution
   stage2-install-distribution
+  clang
+  lld
   CACHE STRING "")
 set(BOOTSTRAP_CLANG_BOOTSTRAP_TARGETS
   clang-bolt
